src/BuiltInTools/dotnet-watch/CommandLine/CommandLineOptions.cs

```diff
@@ -173,12 +192,36 @@ internal sealed class CommandLineOptions
         };
     }
 
+    /// <summary>
+    /// Parses the value of msbuild option `-binaryLogger[:[LogFile=]output.binlog[;ProjectImports={None,Embed,ZipFile}]]`.
+    /// Emulates https://github.com/dotnet/msbuild/blob/7f69ea906c29f2478cc05423484ad185de66e124/src/Build/Logging/BinaryLogger/BinaryLogger.cs#L481.
+    /// See https://github.com/dotnet/msbuild/issues/12256
+    /// </summary>
+    internal static string? ParseBinaryLogFilePath(string? value)
+        => value switch
+        {
+            null => null,
+            _ => (from parameter in value.Split(';', StringSplitOptions.RemoveEmptyEntries)
+                  where !string.Equals(parameter, "ProjectImports=None", StringComparison.OrdinalIgnoreCase) &&
+                        !string.Equals(parameter, "ProjectImports=Embed", StringComparison.OrdinalIgnoreCase) &&
+                        !string.Equals(parameter, "ProjectImports=ZipFile", StringComparison.OrdinalIgnoreCase) &&
+                        !string.Equals(parameter, "OmitInitialInfo", StringComparison.OrdinalIgnoreCase)
+                  let path = (parameter.StartsWith("LogFile=", StringComparison.OrdinalIgnoreCase) ? parameter["LogFile=".Length..] : parameter).Trim('"')
+                  let pathWithExtension = path.EndsWith(".binlog", StringComparison.OrdinalIgnoreCase) ? path : $"{path}.binlog"
+                  select pathWithExtension)
+                 .LastOrDefault("msbuild.binlog")
+        };
+
     private static IReadOnlyList<string> GetCommandArguments(
```

Comment:
Future thought: Should this return some sort of object that includes the command arguments, the binlog arguments, and other (future) pieces?

Reply:
Binlog shouldn't be parsed this way. This is just a workaround for https://github.com/dotnet/sdk/issues/49989.